File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/JsonParser.java
package org.hyperledger.iroha.android.client;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Minimal JSON parser sufficient for the SDK polling helpers.
 *
 * <p>Only the subset required by SDK payloads is implemented (objects, arrays, strings, booleans,
 * null, numbers).
 */
public final class JsonParser {

  private static final int MAX_NESTING_DEPTH = 128;

  private final String input;
  private int index;

  private JsonParser(final String input) {
    this.input = input;
  }

  public static Object parse(final String json) {
    final JsonParser parser = new JsonParser(json);
    parser.skipWhitespace();
    final Object value = parser.parseValue(0);
    parser.skipWhitespace();
    if (parser.index != parser.input.length()) {
      throw new IllegalStateException("Trailing characters after JSON payload");
    }
    return value;
  }

  private Object parseValue(final int depth) {
    if (depth > MAX_NESTING_DEPTH) {
      throw new IllegalStateException("JSON exceeds maximum nesting depth");
    }
    skipWhitespace();
    if (index >= input.length()) {
      throw new IllegalStateException("Unexpected end of JSON input");
    }
    final char c = input.charAt(index);
    return switch (c) {
      case '{' -> parseObject(depth);
      case '[' -> parseArray(depth);
      case '"' -> parseString();
      case 't' -> {
        consumeLiteral("true");
        yield Boolean.TRUE;
      }
      case 'f' -> {
        consumeLiteral("false");
        yield Boolean.FALSE;
      }
      case 'n' -> {
        consumeLiteral("null");
        yield null;
      }
      default -> parseNumber();
    };
  }

  private Map<String, Object> parseObject(final int depth) {
    expect('{');
    skipWhitespace();
    final Map<String, Object> map = new LinkedHashMap<>();
    if (peek('}')) {
      index++;
      return map;
    }
    while (true) {
      final String key = parseString();
      if (map.containsKey(key)) {
        throw new IllegalStateException("Duplicate JSON object key: " + key);
      }
      skipWhitespace();
      expect(':');
      skipWhitespace();
      map.put(key, parseValue(depth + 1));
      skipWhitespace();
      if (peek('}')) {
        index++;
        return map;
      }
      expect(',');
      skipWhitespace();
    }
  }

  private List<Object> parseArray(final int depth) {
    expect('[');
    skipWhitespace();
    final List<Object> list = new ArrayList<>();
    if (peek(']')) {
      index++;
      return list;
    }
    while (true) {
      list.add(parseValue(depth + 1));
      skipWhitespace();
      if (peek(']')) {
        index++;
        return list;
      }
      expect(',');
      skipWhitespace();
    }
  }

  private String parseString() {
    expect('"');
    final StringBuilder builder = new StringBuilder();
    while (index < input.length()) {
      final char c = input.charAt(index++);
      if (c == '"') {
        return builder.toString();
      }
      if (c == '\\') {
        if (index >= input.length()) {
          throw new IllegalStateException("Invalid escape sequence");
        }
        final char esc = input.charAt(index++);
        switch (esc) {
          case '"' -> builder.append('"');
          case '\\' -> builder.append('\\');
          case '/' -> builder.append('/');
          case 'b' -> builder.append('\b');
          case 'f' -> builder.append('\f');
          case 'n' -> builder.append('\n');
          case 'r' -> builder.append('\r');
          case 't' -> builder.append('\t');
          case 'u' -> {
            final char high = parseUnicodeEscapeUnit();
            if (Character.isHighSurrogate(high)) {
              if (index + 2 > input.length()
                  || input.charAt(index) != '\\'
                  || input.charAt(index + 1) != 'u') {
                throw new IllegalStateException("Invalid unicode surrogate pair");
              }
              index += 2;
              final char low = parseUnicodeEscapeUnit();
              if (!Character.isLowSurrogate(low)) {
                throw new IllegalStateException("Invalid unicode surrogate pair");
              }
              builder.append(high).append(low);
            } else if (Character.isLowSurrogate(high)) {
              throw new IllegalStateException("Invalid unicode surrogate pair");
            } else {
              builder.append(high);
            }
          }
          default -> throw new IllegalStateException("Unsupported escape: \\" + esc);
        }
      } else {
        if (c < 0x20) {
          throw new IllegalStateException("Unescaped control character in JSON string");
        }
        if (Character.isHighSurrogate(c)) {
          if (index >= input.length() || !Character.isLowSurrogate(input.charAt(index))) {
            throw new IllegalStateException("Invalid unicode surrogate pair");
          }
          builder.append(c).append(input.charAt(index++));
        } else if (Character.isLowSurrogate(c)) {
          throw new IllegalStateException("Invalid unicode surrogate pair");
        } else {
          builder.append(c);
        }
      }
    }
    throw new IllegalStateException("Unterminated string literal");
  }

  private char parseUnicodeEscapeUnit() {
    if (index + 4 > input.length()) {
      throw new IllegalStateException("Invalid unicode escape");
    }
    final String hex = input.substring(index, index + 4);
    index += 4;
    try {
      return (char) Integer.parseInt(hex, 16);
    } catch (NumberFormatException error) {
      throw new IllegalStateException("Invalid unicode escape", error);
    }
  }

  private Number parseNumber() {
    final int start = index;
    if (index < input.length() && input.charAt(index) == '-') {
      index++;
    }
    if (index >= input.length()) {
      throw new IllegalStateException("Invalid number: expected digit");
    }
    boolean hasDigits = false;
    if (index < input.length() && isDigit(input.charAt(index))) {
      hasDigits = true;
      if (input.charAt(index) == '0') {
        index++;
        if (index < input.length() && isDigit(input.charAt(index))) {
          throw new IllegalStateException("Invalid number: leading zero");
        }
      } else {
        while (index < input.length() && isDigit(input.charAt(index))) {
          index++;
        }
      }
    }
    if (!hasDigits) {
      throw new IllegalStateException("Invalid number: expected digit");
    }
    boolean hasFraction = false;
    if (index < input.length() && input.charAt(index) == '.') {
      hasFraction = true;
      index++;
      if (index >= input.length() || !isDigit(input.charAt(index))) {
        throw new IllegalStateException("Invalid number: missing digit after decimal point");
      }
      while (index < input.length() && isDigit(input.charAt(index))) {
        index++;
      }
    }
    boolean hasExponent = false;
    if (index < input.length()) {
      final char exp = input.charAt(index);
      if (exp == 'e' || exp == 'E') {
        hasExponent = true;
        index++;
        if (index < input.length()) {
          final char sign = input.charAt(index);
          if (sign == '+' || sign == '-') {
            index++;
          }
        }
        if (index >= input.length() || !isDigit(input.charAt(index))) {
          throw new IllegalStateException("Invalid number: missing exponent digits");
        }
        while (index < input.length() && isDigit(input.charAt(index))) {
          index++;
        }
      }
    }
    final String token = input.substring(start, index);
    try {
      if (!hasFraction && !hasExponent) {
        try {
          return Long.parseLong(token);
        } catch (NumberFormatException ex) {
          return new BigInteger(token);
        }
      }
      return new BigDecimal(token);
    } catch (NumberFormatException ex) {
      throw new IllegalStateException("Invalid number: " + token, ex);
    }
  }

  private void consumeLiteral(final String literal) {
    if (!input.regionMatches(index, literal, 0, literal.length())) {
      throw new IllegalStateException("Expected literal '" + literal + "'");
    }
    index += literal.length();
  }

  private void skipWhitespace() {
    while (index < input.length()) {
      final char c = input.charAt(index);
      if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
        index++;
      } else {
        break;
      }
    }
  }

  private void expect(final char expected) {
    if (index >= input.length() || input.charAt(index) != expected) {
      throw new IllegalStateException("Expected '" + expected + "'");
    }
    index++;
  }

  private boolean peek(final char expected) {
    return index < input.length() && input.charAt(index) == expected;
  }

  private static boolean isDigit(final char c) {
    return c >= '0' && c <= '9';
  }
}

File: .taira-source-c80bf4d85a97/kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/JsonParser.kt
package org.hyperledger.iroha.sdk.client

/**
 * Minimal JSON parser sufficient for the SDK polling helpers.
 *
 * Only the subset required by SDK payloads is implemented (objects, arrays, strings, booleans,
 * null, numbers).
 */
class JsonParser private constructor(private val input: String) {

    private var index = 0

    private fun parseValue(depth: Int): Any? {
        check(depth <= MAX_NESTING_DEPTH) { "JSON exceeds maximum nesting depth" }
        skipWhitespace()
        check(index < input.length) { "Unexpected end of JSON input" }
        return when (input[index]) {
            '{' -> parseObject(depth)
            '[' -> parseArray(depth)
            '"' -> parseString()
            't' -> { consumeLiteral("true"); true }
            'f' -> { consumeLiteral("false"); false }
            'n' -> { consumeLiteral("null"); null }
            else -> parseNumber()
        }
    }

    private fun parseObject(depth: Int): LinkedHashMap<String, Any?> {
        expect('{')
        skipWhitespace()
        val map = LinkedHashMap<String, Any?>()
        if (peek('}')) { index++; return map }
        while (true) {
            val key = parseString()
            check(!map.containsKey(key)) { "Duplicate JSON object key: $key" }
            skipWhitespace()
            expect(':')
            skipWhitespace()
            map[key] = parseValue(depth + 1)
            skipWhitespace()
            if (peek('}')) { index++; return map }
            expect(',')
            skipWhitespace()
        }
    }

    private fun parseArray(depth: Int): MutableList<Any?> {
        expect('[')
        skipWhitespace()
        val list = mutableListOf<Any?>()
        if (peek(']')) { index++; return list }
        while (true) {
            list.add(parseValue(depth + 1))
            skipWhitespace()
            if (peek(']')) { index++; return list }
            expect(',')
            skipWhitespace()
        }
    }

    private fun parseString(): String {
        expect('"')
        val builder = StringBuilder()
        while (index < input.length) {
            val c = input[index++]
            if (c == '"') return builder.toString()
            if (c == '\\') {
                check(index < input.length) { "Invalid escape sequence" }
                when (val esc = input[index++]) {
                    '"' -> builder.append('"')
                    '\\' -> builder.append('\\')
                    '/' -> builder.append('/')
                    'b' -> builder.append('\b')
                    'f' -> builder.append('\u000C')
                    'n' -> builder.append('\n')
                    'r' -> builder.append('\r')
                    't' -> builder.append('\t')
                    'u' -> {
                        val high = parseUnicodeEscapeUnit()
                        if (Character.isHighSurrogate(high)) {
                            check(index + 2 <= input.length && input[index] == '\\' && input[index + 1] == 'u') {
                                "Invalid unicode surrogate pair"
                            }
                            index += 2
                            val low = parseUnicodeEscapeUnit()
                            check(Character.isLowSurrogate(low)) { "Invalid unicode surrogate pair" }
                            builder.append(high).append(low)
                        } else {
                            check(!Character.isLowSurrogate(high)) { "Invalid unicode surrogate pair" }
                            builder.append(high)
                        }
                    }
                    else -> throw IllegalStateException("Unsupported escape: \\$esc")
                }
            } else {
                check(c.code >= 0x20) { "Unescaped control character in JSON string" }
                when {
                    Character.isHighSurrogate(c) -> {
                        check(index < input.length && Character.isLowSurrogate(input[index])) {
                            "Invalid unicode surrogate pair"
                        }
                        builder.append(c).append(input[index])
                        index++
                    }
                    Character.isLowSurrogate(c) ->
                        throw IllegalStateException("Invalid unicode surrogate pair")
                    else -> builder.append(c)
                }
            }
        }
        throw IllegalStateException("Unterminated string literal")
    }

    private fun parseUnicodeEscapeUnit(): Char {
        check(index + 4 <= input.length) { "Invalid unicode escape" }
        val value = input.substring(index, index + 4).toIntOrNull(16)
        check(value != null) { "Invalid unicode escape" }
        index += 4
        return value.toChar()
    }

    private fun parseNumber(): Number {
        val start = index
        if (index < input.length && input[index] == '-') index++
        check(index < input.length) { "Invalid number: expected digit" }
        var hasDigits = false
        if (index < input.length && input[index].isDigit()) {
            hasDigits = true
            if (input[index] == '0') {
                index++
                check(index >= input.length || !input[index].isDigit()) { "Invalid number: leading zero" }
            } else {
                while (index < input.length && input[index].isDigit()) index++
            }
        }
        check(hasDigits) { "Invalid number: expected digit" }
        var hasFraction = false
        if (index < input.length && input[index] == '.') {
            hasFraction = true
            index++
            check(index < input.length && input[index].isDigit()) { "Invalid number: missing digit after decimal point" }
            while (index < input.length && input[index].isDigit()) index++
        }
        var hasExponent = false
        if (index < input.length && (input[index] == 'e' || input[index] == 'E')) {
            hasExponent = true
            index++
            if (index < input.length && (input[index] == '+' || input[index] == '-')) index++
            check(index < input.length && input[index].isDigit()) { "Invalid number: missing exponent digits" }
            while (index < input.length && input[index].isDigit()) index++
        }
        val token = input.substring(start, index)
        return try {
            if (!hasFraction && !hasExponent) {
                try {
                    token.toLong()
                } catch (_: NumberFormatException) {
                    java.math.BigInteger(token)
                }
            } else {
                java.math.BigDecimal(token)
            }
        } catch (ex: NumberFormatException) {
            throw IllegalStateException("Invalid number: $token", ex)
        }
    }

    private fun consumeLiteral(literal: String) {
        check(input.regionMatches(index, literal, 0, literal.length)) { "Expected literal '$literal'" }
        index += literal.length
    }

    private fun skipWhitespace() {
        while (index < input.length && input[index] in " \t\n\r") index++
    }

    private fun expect(expected: Char) {
        check(index < input.length && input[index] == expected) { "Expected '$expected'" }
        index++
    }

    private fun peek(expected: Char): Boolean =
        index < input.length && input[index] == expected

    private fun Char.isDigit(): Boolean = this in '0'..'9'

    companion object {
        private const val MAX_NESTING_DEPTH = 128

        @JvmStatic
        fun parse(json: String): Any? {
            val parser = JsonParser(json)
            parser.skipWhitespace()
            val value = parser.parseValue(0)
            parser.skipWhitespace()
            check(parser.index == parser.input.length) { "Trailing characters after JSON payload" }
            return value
        }
    }
}
